test: create assert_not_equal util

kevkevinpal opened the pull request:
In the functional tests there are lots of cases where we assert != which we now swap with assert_not_equal to be more readable

This is motivated/uses logic from this PR which was closed https://github.com/bitcoin/bitcoin/pull/28528
This partially helps https://github.com/bitcoin/bitcoin/issues/23119

I've broken it up to just `assert_not_equal` to keep the PR smaller as suggested in https://github.com/bitcoin/bitcoin/pull/28528#issuecomment-1959945805

I can create follow up PR's if this is wanted

kevkevinpal committed:
test: create assert_not_equal util and add to where imports are needed

In the functional tests there are lots of cases where we assert != which
this new util will replace, we also are adding the imports and the new assertion

axizqayum686 reviewed (changes requested):
.

Empact commented:
Would be nice to structure this as a [scripted diff](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#scripted-diffs): one commit where you add the function and requires, another where you convert the `assert !=`s.

kevkevinpal commented:
added scripted diff in [005a8e9](https://github.com/bitcoin/bitcoin/pull/29500/commits/005a8e9211f3988154b9e5d0c406f4c57c4d0a96)

with the following script
`git grep -l "assert.*!=" ./test/functional`

I will have to amend the commit as I see more spots I need to modify with the `assert_not_equal` util

brunoerg commented on test/functional/p2p_blockfilters.py:
```diff
@@ -21,6 +22,7 @@
 from test_framework.p2p import P2PInterface
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import (
+    assert_not_equal,
```
Importing twice? (from `util` and `messages`)

kevkevinpal replied:
hmm yea I was trying to use this command to find all the places but it seems to be a bit more complex than I initially thought
```
git grep -l "assert .*!=" -- ':(exclude)*script.py' ./test/functional | xargs sed -i -e "/assert .*!=/ s/assert /assert_not_equal(/g" -e "/assert_not_equal/ s/ !=/,/g" -e "/assert_not_equal/ s/$/)/g" -e "s/test_framework.util import (/test_framework.util import (\n    assert_not_equal,/g" -e "s/import assert_equal/(\n    assert_equal,\n    assert_not_equal,\n)/g"
```

I might just manually add the imports and use the script to look for the `assert ... !=` as suggested by @Empact

fanquake commented:
https://cirrus-ci.com/task/6627384883937280?logs=ci#L4021
```bash
 test  2024-03-05T05:30:24.555000Z TestFramework (ERROR): Unexpected exception caught during testing 
                                   Traceback (most recent call last):
                                     File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 131, in main
                                       self.run_test()
                                     File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/mempool_accept_v3.py", line 448, in run_test
                                       self.test_reorg_2child_rbf()
                                     File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/mempool_accept_v3.py", line 29, in wrapper
                                       func(self)
                                     File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/mempool_accept_v3.py", line 429, in test_reorg_2child_rbf
                                       assert_not_equal((child_1_conflict["txid"], child_1["txid"]))
                                   TypeError: assert_not_equal() missing 1 required positional argument: 'thing2'
```

kevkevinpal commented:
not sure why this ci is failing `ASan + LSan + UBSan + integer, no depends, USDT` but the rest seem to be passing

kevkevinpal commented:
> > not sure why this ci is failing `ASan + LSan + UBSan + integer, no depends, USDT` but the rest seem to be passing
> 
> It is a race in the test logic between the python code and the `net` thread, IIUC. Please file a new issue or a new pull to fix it.

Opened a new issue for the CI error

BrandonOdiwuor reviewed (commented):
Approach ACK

Good job approaching this using Scripted diffs

rkrux reviewed (commented):
crACK [f9ba6fd](https://github.com/bitcoin/bitcoin/pull/29500/commits/f9ba6fd46d454d0a56662db751675a6520e1c9f6)

I'm unable to run make successfully because of the following error on my system. I've noticed this in another PR as well and it is due to lack of a commit that was merged in later. 
Asked couple questions.

```
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/time.h:127:12: note: 'gmtime_r' declared here
struct tm *gmtime_r(const time_t * __restrict, struct tm * __restrict);
```

rkrux commented:
@kevkevinpal Did you use a tool to generate the verification script?

ajtowns commented on test/functional/feature_dbcrash.py:
```diff
@@ -274,7 +275,7 @@ def run_test(self):
 
         # If no nodes were restarted, we didn't test anything.
-        assert self.restart_counts != [0, 0, 0]
+        assert_not_equal(self.restart_counts, [0, 0, 0])
```
`assert_equal` is helpful because when `a == 3` fails, it can be helpful to know what the value of `a` actually is without having to add debug code and rerun. If `a != 3` fails, though, you already know that `a == 3`, so this doesn't seem very helpful.

There are cases where you're comparing `a != b` and may not be sure which ones getting the wrong value where this could perhaps be helpful, but those seem pretty rare.

Concept -0 for me.

vasild replied:
> There are cases where you're comparing a != b and may not be sure which ones getting the wrong value where this could perhaps be helpful, but those **seem pretty rare**

I counted (by eyes, could be +/- a few):
* 16 occurrences where one of the arguments is a constant, e.g. `assert_not_equal(peer.nServices & NODE_BLOOM, 0)`, and
* 59 occurrences where both arguments are variables, e.g. `assert_not_equal(child_one_wtxid, child_two_wtxid)`

ajtowns commented on test/functional/test_framework/util.py:
```diff
@@ -56,6 +56,10 @@ def assert_equal(thing1, thing2, *args):
 
+def assert_not_equal(thing1, thing2, *args):
+    if thing1 == thing2 or any(thing1 == arg for arg in args):
+        raise AssertionError("%s" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
```
This error message doesn't make sense when `*args` is non-empty: if you say `assert_not_equal(a, 1, 2)` and `a` is 1 or 2, it will report `1 == 1 == 2` or `2 == 1 == 2`.

kevkevinpal replied:
I've updated this to instead only take three arguments and the last one being an optional message since I did not see anywhere we were already asserting `!=` in a chain.

This should clear up the message and allow the one location we do use a message to be used

updated in [4488120](https://github.com/bitcoin/bitcoin/pull/29500/commits/4488120ab7db775ce0ae8a73f910dddca1dee123)

kevkevinpal commented on test/functional/test_framework/test_node.py:
```diff
@@ -638,7 +638,7 @@ def assert_start_raises_init_error(self, extra_args=None, expected_msg=None, mat
                 ret = self.process.wait(timeout=self.rpc_timeout)
                 self.log.debug(self._node_msg(f'bitcoind exited with status {ret} during initialization'))
-                assert ret != 0  # Exit code must indicate failure
+                assert_not_equal(ret, 0  ) # Exit code must indicate failure
```
not sure why it is doing that I will check the script to see why

kevkevinpal replied:
fixed in [4358c6e](https://github.com/bitcoin/bitcoin/pull/29500/commits/4358c6e7221eed21d9cf78ead5cabd6982fb9a03)

thanks for the review!

kevkevinpal commented on test/functional/p2p_blockfilters.py:
```diff
@@ -94,7 +94,7 @@ def run_test(self):
 
         main_block_hash = self.nodes[0].getblockhash(1000)
-        assert main_block_hash != stale_block_hash, "node 0 chain did not reorganize"
+        assert_not_equal(main_block_hash, stale_block_hash, "node 0 chain did not reorganize")
```
that is strange I would think that the 3rd arg would make the tests fail in this case

kevkevinpal replied:
I've updated the function to not accept the `args` param since it did not seem to be needed and instead allowed an optional 3rd param as the message

updated in [4488120](https://github.com/bitcoin/bitcoin/pull/29500/commits/4488120ab7db775ce0ae8a73f910dddca1dee123)

kevkevinpal commented:
> @kevkevinpal Did you use a tool to generate the verification script?

No I did not I just manually used `git grep` and `sed` to create the script

kevkevinpal commented:
rebased to [7f7dc43](https://github.com/bitcoin/bitcoin/pull/29500/commits/7f7dc439c7d268e63eddddf3cea21c2931895078)

ceffikhan commented:
> follow



> In the functional tests there are lots of cases where we assert != which we now swap with assert_not_equal to be more readable
> 
> This is motivated/uses logic from this PR which was closed #28528 This partially helps #23119
> 
> I've broken it up to just `assert_not_equal` to keep the PR smaller as suggested in [#28528 (comment)](https://github.com/bitcoin/bitcoin/pull/28528#issuecomment-1959945805)
> 
> I can create follow up PR's if this is wanted

cdcd

ceffikhan commented:
> In the functional tests there are lots of cases where we assert != which we now swap with assert_not_equal to be more readable
> 
> This is motivated/uses logic from this PR which was closed #28528 This partially helps #23119
> 
> I've broken it up to just `assert_not_equal` to keep the PR smaller as suggested in [#28528 (comment)](https://github.com/bitcoin/bitcoin/pull/28528#issuecomment-1959945805)
> 
> I can create follow up PR's if this is wanted

I also had the motivation to do this kind of stuff with other symbols like greater than and less than I posted this motivation on irc but came to the conclusion that plain != is better for non english speakers and for english speakers readability remains the same.

glozow commented:
Are you still working on this?

kevkevinpal commented:
> Are you still working on this?

Yup just haven't had much activity on it but rebased to [1d72266](https://github.com/bitcoin/bitcoin/pull/29500/commits/1d722660a65522539872c09ae8a3ba8c9ca55b77)

vasild reviewed (approved):
Almost ACK 1d722660a6, modulo squash of the two commits.

This is a nice addition.

It is better to squash the two commits into one because IMO it does not make sense to first add the import in one commit and in another commit to use it. That way it becomes difficult to review whether an unnecessary import was added to some file. Also some python linter that tests each commit could be upset about the unused imports in the first commit.

I found it easier to read the diff itself instead of the "scripted diff" code (the line is 700+ chars long!). Further, I tweaked the diff to get the left and right sides align, so instead of:

```diff
-        assert (cblks[3]['0'][33:65] < cblks[3]['1'][33:65]) != (cblks[4]['0'][33:65] < cblks[4]['1'][33:65])
+        assert_not_equal((cblks[3]['0'][33:65] < cblks[3]['1'][33:65]), (cblks[4]['0'][33:65] < cblks[4]['1'][33:65]))
```

I looked at:

```diff
-        assert (cblks[3]['0'][33:65] < cblks[3]['1'][33:65]) != (cblks[4]['0'][33:65] < cblks[4]['1'][33:65])
+        ass_ne((cblks[3]['0'][33:65] < cblks[3]['1'][33:65]),   (cblks[4]['0'][33:65] < cblks[4]['1'][33:65]))
```
I used `git show 1d722660a6 |sed -E 's/^(.*)assert_not_equal([^,]*), (.*)$/\1ass_ne\2,   \3/g'` for that.

Needs rebase for CMake.

rkrux commented on test/functional/test_framework/util.py:
```diff
@@ -74,6 +74,10 @@ def assert_equal(thing1, thing2, *args):
     if thing1 != thing2 or any(thing1 != arg for arg in args):
         raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
 
+def assert_not_equal(thing1, thing2, message=""):
```
Nit: The 3rd argument can be called `error_message`.

rkrux reviewed (commented):
Concept ACK 1d722660a65522539872c09ae8a3ba8c9ca55b77
The util function looks quite different from the last time I reviewed, simpler and more readable now!

Couple points:
1. The PR needs a rebase so that it works with the new build system, unable to build and test it in local atm with the latest build commands. Though the PR doesn't have any conflicts with master, it would be nice to have this PR rebased if not much overhead for the author.
2. I agree with @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/29500#pullrequestreview-2415263506) that the first commit should not be importing the helper function only to be used in the next commit. While reviewing the first commit, it immediately struck to me it is being imported but not used, which looked odd. However, I now wonder would squashing the 2 commits still work with the scripted diff as per the comment in the script checker?

https://github.com/bitcoin/bitcoin/blob/master/test/lint/commit-script-check.sh#L11-L12

> The resulting script should exactly transform the previous commit into the current
one. Any remaining diff signals an error.

kevkevinpal commented:
rebased to [6590b26](https://github.com/bitcoin/bitcoin/pull/29500/commits/6590b267015ad094dfcbacad521c30207fe199bd)

and I can address your latest comments soon

kevkevinpal commented:
yes @rkrux you are right if I want to use a scripted diff then all the changes must be done using the script

I can remove the scripted diff and just squash them into a single diff since right now the script is getting very large already if that is preferred by others

fanquake commented:
@kevkevinpal what's the status of this?

vasild commented:
Almost ACK dcf9a45aab4c4a7f9f900d5dafcdfe0e9a598a38, modulo squash the two commits into one and rebase.

kevkevinpal commented:
> @kevkevinpal what's the status of this?

Just rebased and addressed @vasild comment

---

> Almost ACK [dcf9a45](https://github.com/bitcoin/bitcoin/commit/dcf9a45aab4c4a7f9f900d5dafcdfe0e9a598a38), modulo squash the two commits into one and rebase.

Just rebased and squashed the commits, I removed the scripted-diff because it was getting more difficult to review the diff than the actual code. But now all the changes are in [d9cb032](https://github.com/bitcoin/bitcoin/pull/29500/commits/d9cb032ec1aa7000fc3f5b9fa48405f269879497)

---

if you do `grep -nr assert\ .*!= ./test/functional/` there should only be one result 

`assert len(ss) == 175 - (in_type == SIGHASH_ANYONECANPAY) * 49 - (out_type != SIGHASH_ALL and out_type != SIGHASH_SINGLE) * 32 + (annex is not None) * 32 + scriptpath * 37`

which we don't need to add `assert_not_equal` to

hodlinator reviewed (commented):
Concept ACK d9cb032ec1aa7000fc3f5b9fa48405f269879497

vasild reviewed (approved):
ACK d9cb032ec1aa7000fc3f5b9fa48405f269879497

ryanofsky reviewed (approved):
Code review ACK d9cb032ec1aa7000fc3f5b9fa48405f269879497. Seems like a nice change, though needs to be rebased. I think this change is good because it makes tests more readable, prints more information on failures, avoids misusing the `assert` keyword, and makes the test framework api more consistent internally and consistent with other frameworks.

kevkevinpal commented:
rebased to [29fa2b9](https://github.com/bitcoin/bitcoin/pull/29500/commits/29fa2b98a8988f9525773ae4b4994ea7f2d64776)

hodlinator commented on test/functional/test_framework/util.py:
```diff
@@ -76,6 +76,10 @@ def assert_equal(thing1, thing2, *args):
 
+def assert_not_equal(thing1, thing2, error_message=""):
+    if thing1 == thing2:
+        raise AssertionError("%s == %s %s" % (str(thing1), str(thing2), str(error_message)))
```
Would be more eager to ACK if code introduced by the PR followed bitcoin/test/functional/README.md as noted above. If you don't want to change the phrasing towards what's in the nit-part that's fine.

kevkevinpal replied:
Went ahead an used your suggestion in [be71af3](https://github.com/bitcoin/bitcoin/pull/29500/commits/be71af3cc0b0bcb7d917cc6f2e5fda119f1b1bd6)

it now looks like this
```
2025-03-27T23:14:10.629000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/mnt/shared_drive/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 182, in main
    self.run_test()
  File "/mnt/shared_drive/DEVDIR/bitcoin/./build_dir/test/functional/p2p_blockfilters.py", line 98, in run_test
    assert_not_equal(stale_block_hash, stale_block_hash, "node 0 chain did not reorganize")
  File "/mnt/shared_drive/DEVDIR/bitcoin/test/functional/test_framework/util.py", line 81, in assert_not_equal
    raise AssertionError(f"Both values are {thing1}, {f'{error_message}' if error_message else ''}")
AssertionError: Both values are 05ba865bb920e2ff5763411a9dd3c225ace8ff4393469bb25acffd56df1bfb6e, node 0 chain did not reorganize
```

hodlinator reviewed (approved):
ACK be71af3cc0b0bcb7d917cc6f2e5fda119f1b1bd6

This change adds the missing opposite of `assert_equal()` and decreases usage of built-in `assert` for performing always-on checks during tests.

hodlinator commented on test/functional/test_framework/util.py:
```diff
@@ -76,6 +76,10 @@ def assert_equal(thing1, thing2, *args):
     if thing1 != thing2 or any(thing1 != arg for arg in args):
         raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
 
+def assert_not_equal(thing1, thing2, error_message=""):
+    if thing1 == thing2:
+        raise AssertionError(f"Both values are {thing1}, {f'{error_message}' if error_message else ''}")
```
> "AssertionError: Both values are (the same)"

I don't know if you mean literally `"Both values are (the same)"` or `f"Both values are {thing1}"`. I find the latter clearly more useful; providing clues of cause of the failure, making it easier to reproduce.

---

What it comes down to with custom error messages is being able to directly communicate with whoever is troubleshooting a failure. If one doesn't even need to go to the logs to see what happened before, I consider it a small win. But agree it is bikeshedding-prone, so I would be okay with removing it.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29500#discussion_r2018038275

I do like error_message parameter idea, fwiw. I could see these making tests more readable if they started being used for more important or more confusing checks. Tests are often written with some important checks directly related to the thing being tested, and other less important checks to see if other things are consistent. Having a place to indicate what it actually means when a check fails could be helpful for more important checks to distinguish them.

yancyribbens replied:
> The traceback in this case does print the source code line, so a comment on the same line could be sufficient. It depends on how the error is presented. Example failure:

This is an annoying draw back that the stack trace won't contain the line in the test that's in error, but instead the line in the `assert_not_equal` helper.  In Rust, one can decorate the helper function with the `track_caller`  macro.  I don't see anything that is similar for C++ sadly.

hodlinator replied:
We're in Python here though. I spent a couple of hours with an LLM in mid-February to brute force this embryo:
https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:2025/02/assert_ergonomics

<details><summary>Output examples</summary>

Current `assert_equal()`-behavior:
```
2025-03-28T21:15:36.379000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 178, in main
    self.run_test()
  File "/home/hodlinator/bitcoin/./build/test/functional/wallet_multisig_descriptor_psbt.py", line 80, in run_test
    assert_equal(
  File "/home/hodlinator/bitcoin/test/functional/test_framework/util.py", line 77, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(0 == 123)
```
With first commit, chomping off util.py traceback entry you complained about:
```
2025-03-28T21:16:19.814000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 178, in main
    self.run_test()
  File "/home/hodlinator/bitcoin/./build/test/functional/wallet_multisig_descriptor_psbt.py", line 80, in run_test
    assert_equal(
AssertionError: not(0 == 123)
```
With second commit, replacing another traceback entry with proper context:
```
2025-03-28T21:17:39.494000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 178, in main
    self.run_test()
AssertionError: not(0 == 123)
Failing expression: /home/hodlinator/bitcoin/./build/test/functional/wallet_multisig_descriptor_psbt.py:80 - 83
        assert_equal(
            123 if (False and
            True) else 0,
            123)
```
</details>

The code needs more work, should be more defensive, forgot to output function name (run_test). And I wouldn't bet on it being accepted anyway.

rkrux replied:
I don't mind having an error message being logged and I do agree with the following points:

> It is not a named arg, nor type-safe, so someone writing assert_not_equal("a", "b", "c") or assert_not_equal(1, 2, 3) will be wrong and confusing at best.

> Unique and descriptive error messages make sense when it is the only piece of information given (for example in the context of an init error during startup of the program).

As I see from the usages of `assert_not_equal`, there is only 1 case where the error message is passed out of ~77 total usages. I believe a named argument for the error message could help in avoiding unintended usages of this utility function.

janb84 reviewed (commented):
Concept ACK [be71af3](https://github.com/bitcoin/bitcoin/commit/be71af3cc0b0bcb7d917cc6f2e5fda119f1b1bd6)

- Code review ✅
- Build & tested ✅

I agree with @hodlinator on the import order NIT — let's keep it consistent and maintain clean, high-quality code. https://github.com/bitcoin/bitcoin/pull/29500#discussion_r1967667753

ryanofsky reviewed (approved):
Code review ACK be71af3cc0b0bcb7d917cc6f2e5fda119f1b1bd6. Just rebased and added f-string since last review.

rkrux reviewed (changes requested):
Concept ACK be71af3cc0b0bcb7d917cc6f2e5fda119f1b1bd6

Requesting changes mainly because an additional comma is printed in case of an assertion error when the error message is not passed, which can be confusing for the reader later.

```
2025-03-31T07:28:44.958000Z TestFramework (INFO): Test fundrawtxn with locked wallet and hardened derivation
2025-03-31T07:28:46.455000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 182, in main
    self.run_test()
    ~~~~~~~~~~~~~^^
  File "/Users/rkrux/projects/bitcoin/./build/test/functional/wallet_fundrawtransaction.py", line 141, in run_test
    self.test_locked_wallet()
    ~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/rkrux/projects/bitcoin/./build/test/functional/wallet_fundrawtransaction.py", line 665, in test_locked_wallet
    assert_not_equal(fundedTx["changepos"], 1)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/util.py", line 81, in assert_not_equal
    raise AssertionError(f"Both values are {thing1}, {f'{error_message}' if error_message else ''}")
AssertionError: Both values are 1, 
2025-03-31T07:28:46.512000Z TestFramework (INFO): Not stopping nodes as test failed. The dangling processes will be cleaned up later.
```

kevkevinpal commented:
Updated to [f3d4b6d](https://github.com/bitcoin/bitcoin/pull/29500/commits/f3d4b6d9627950240f97bcb9d18bb415154b4fec)

Added named variable for `error_message` and also made the assert able to take more than two variables to compare to each other. The error will now say 
`Two or more values equal to <value>: <error_message>` or
`Two or more values equal to <value>`

if no `error_message` provided

kevkevinpal commented on test/functional/test_framework/util.py:
```diff
@@ -76,6 +76,16 @@ def assert_equal(thing1, thing2, *args):
+            if val1 == val2:
+                raise AssertionError(
+                    f"Two or more values equal to {val1}{f': {error_message}' if error_message else ''}"
+                )
```
fair point and thanks for the review

I've updated to [7bb83f6](https://github.com/bitcoin/bitcoin/pull/29500/commits/7bb83f6718110cb3a29e72522fdc5515db21c7d0)

- we are now using the error message `(f"Both values are {thing1}{f', {error_message}' if error_message else ''}"`
- we only take two positional arguments and a named `error_message` arg which is used in `p2p_blockfilters`

kevkevinpal commented on test/functional/test_framework/util.py:
```diff
@@ -76,6 +76,16 @@ def assert_equal(thing1, thing2, *args):
     if thing1 != thing2 or any(thing1 != arg for arg in args):
         raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
 
+def assert_not_equal(*things,  error_message=""):
```
resolved in [7bb83f6](https://github.com/bitcoin/bitcoin/pull/29500/commits/7bb83f6718110cb3a29e72522fdc5515db21c7d0)

I went back to using two params and made `error_message` a named parameter that cannot be used as a positional argument

hodlinator reviewed (approved):
re-ACK 7bb83f6718110cb3a29e72522fdc5515db21c7d0

Remaining reservations from  https://github.com/bitcoin/bitcoin/pull/29500#discussion_r2018254263 (sorry for broken record):
* Current push still uses *falsy* `error_message=""` with the `if error_message`, instead of clear `error_message=None`
* `error_message=` at call sites is quite verbose.

ryanofsky reviewed (approved):
Code review ACK 7bb83f6718110cb3a29e72522fdc5515db21c7d0. Only change since last review is fixing error message formatting and passing it as a keyword argument

janb84 reviewed (commented):
Re-ACK [7bb83f6](https://github.com/bitcoin/bitcoin/commit/7bb83f6718110cb3a29e72522fdc5515db21c7d0)

rkrux commented on test/functional/test_framework/test_node.py:
```diff
@@ -668,7 +669,7 @@ def assert_start_raises_init_error(self, extra_args=None, expected_msg=None, mat
                 self.start(extra_args, stdout=log_stdout, stderr=log_stderr, *args, **kwargs)
                 ret = self.process.wait(timeout=self.rpc_timeout)
                 self.log.debug(self._node_msg(f'bitcoind exited with status {ret} during initialization'))
-                assert ret != 0  # Exit code must indicate failure
+                assert_not_equal(ret, 0) # Exit code must indicate failure
```
The error message arg can/could have been used here instead of the comment.

rkrux reviewed (commented):
Post-merge ACK 7bb83f6718110cb3a29e72522fdc5515db21c7d0
